### PR TITLE
usb: divided usb client into smaller modules; provided new api for usbclient - handle several endpoints simultaneously.

### PR DESCRIPTION
**Pull request by HBuczynski**

### CODE REVIEW

File: usb/imx-usbc/desc_manager.c [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+
+static void desc_strInit(usb_desc_list_t *desList, int *localOffset, int strOrder)
+{
+	void *vrtAddr;
+
+	switch (strOrder) {
+		case 0:
+			vrtAddr = desc_common.data->setupMem + *localOffset;
```
Comment (agkaminski) [inline usb/imx-usbc/desc_manager.c]: Arithmetic on void *

File: usb/imx-usbc/desc_manager.c [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+			desc_common.str_0 = VM_2_PHYM(vrtAddr);
+			memcpy(vrtAddr, desList->descriptor, sizeof(usb_string_desc_t));
+			*localOffset += desList->descriptor->bFunctionLength;
+			break;
+
+		case 1:
+			vrtAddr = desc_common.data->setupMem + *localOffset;
```
Comment (agkaminski) [inline usb/imx-usbc/desc_manager.c]: Arithmetic on void *

File: usb/imx-usbc/desc_manager.c [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+			desc_common.str_man = VM_2_PHYM(vrtAddr);
+			memcpy(vrtAddr, desList->descriptor, desList->descriptor->bFunctionLength);
+			*localOffset += desList->descriptor->bFunctionLength;
+			break;
+
+		case 2:
+			vrtAddr = desc_common.data->setupMem + *localOffset;
```
Comment (agkaminski) [inline usb/imx-usbc/desc_manager.c]: Arithmetic on void *

File: usb/imx-usbc/desc_manager.c (hunk 1 above)
Comment (agkaminski) [reply usb/imx-usbc/desc_manager.c]: Also, this line is identical on all cases. You can place it before the swich

File: usb/imx-usbc/desc_manager.c [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+
+	memset(desc_common.data->setupMem, 0, USB_BUFFER_SIZE);
+
+	/* Extract mandatory descriptors to mapped memory */
+	for (; desList != NULL; desList = desList->next) {
+
+		if (localOffset > USB_BUFFER_SIZE)
+			return -ENOMEM	;
+
+		switch(desList->descriptor->bDescriptorType) {
```
Comment (agkaminski) [inline usb/imx-usbc/desc_manager.c]: missing space after switch

File: usb/imx-usbc/desc_manager.c [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+			break;
+	}
+}
+
+
+int desc_init(usb_desc_list_t *desList, usb_common_data_t *usb_data_in, usb_dc_t *dc_in)
+{
+	int localOffset = 0;
+	uint32_t string_desc_count = 0;
+	void *vrtAddr;
```
Comment (agkaminski) [inline usb/imx-usbc/desc_manager.c]: As above

File: usb/imx6ull-usbc/phy.c [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+void *phy_getBase(uint32_t size)
+{
+	return mmap(NULL, size, PROT_WRITE | PROT_READ, MAP_DEVICE, OID_PHYSMEM, USB_BASE_ADDR);
+}
+
+
+uint32_t phy_getIrq(void)
+{
+	return 75;
```
Comment (agkaminski) [inline usb/imx6ull-usbc/phy.c]: You can use irq number enum in <phoenix/arch/imxrt.h>

File: usb/imx-usbc/client.c [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+	if (ctrl_init(&imx_common.data, &imx_common.dc) < 0) {
+		usbclient_cleanData();
+		resourceDestroy(imx_common.dc.irqLock);
+		resourceDestroy(imx_common.dc.irqCond);
+		resourceDestroy(imx_common.dc.endp0Lock);
+		resourceDestroy(imx_common.dc.endp0Cond);
+		return -ENOENT;
+	}
+
```
Comment (kaman) [inline usb/imx-usbc/client.c]: code duplication could be avoided. we can obviously use goto but we may not do it. stil i think 
`int res = EOK;
[...]
res |= mutexCreate(&imx_common.dc.endp0Lock);
res | = condCreate(&imx_common.dc.endp0Cond);
[...]
if (res != EOK) {
		resourceDestroy(imx_common.dc.irqLock);
 		resourceDestroy(imx_common.dc.irqCond);
               [...]
}`
 would be a little bit cleaner.
By no means any code should be changed it is just general idea.

**Pull request merged**